File: app/src/main/java/me/sunhapper/spcharedittool/emoji/EmojiconScrollTabBar.java
package me.sunhapper.spcharedittool.emoji;

import android.content.Context;
import android.support.v4.view.ViewCompat;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.HorizontalScrollView;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;
import java.util.ArrayList;
import java.util.List;
import me.sunhapper.spcharedittool.R;
import me.sunhapper.spcharedittool.util.Utils;

public class EmojiconScrollTabBar extends RelativeLayout {

  private Context context;
  private HorizontalScrollView scrollView;
  private LinearLayout tabContainer;
  private TextView sendBtn;

  private List<ImageView> tabList = new ArrayList<ImageView>();
  private EaseScrollTabBarItemClickListener itemClickListener;
  private OnClickListener sendBtnListener;

  private int tabWidth = 60;

  public EmojiconScrollTabBar(Context context) {
    this(context, null);
  }

  public EmojiconScrollTabBar(Context context, AttributeSet attrs, int defStyle) {
    this(context, attrs);
  }

  public EmojiconScrollTabBar(Context context, AttributeSet attrs) {
    super(context, attrs);
    init(context, attrs);
  }

  private void init(Context context, AttributeSet attrs) {
    this.context = context;
    LayoutInflater.from(context).inflate(R.layout.common_emoj_widget_emojicon_tab_bar, this);

    scrollView =  findViewById(R.id.scroll_view);
    tabContainer =  findViewById(R.id.tab_container);
    sendBtn = findViewById(R.id.send_btn);
    sendBtn.setOnClickListener(new OnClickListener() {
      @Override
      public void onClick(View v) {
        if (sendBtnListener != null) {
          sendBtnListener.onClick(v);
        }
      }
    });
  }

  /**
   * 添加tab
   */
  public void addTab(int icon) {
    View tabView = View.inflate(context, R.layout.common_emoj_scroll_tab_item, null);
    ImageView imageView = (ImageView) tabView.findViewById(R.id.iv_icon);
    imageView.setImageResource(icon);
    LinearLayout.LayoutParams imgParams = new LinearLayout.LayoutParams(
        Utils.dp2px(context, tabWidth), LayoutParams.MATCH_PARENT);
    imageView.setLayoutParams(imgParams);
    tabContainer.addView(tabView);
    tabList.add(imageView);
    final int position = tabList.size() - 1;
    imageView.setOnClickListener(new OnClickListener() {

      @Override
      public void onClick(View v) {
        if (itemClickListener != null) {
          itemClickListener.onItemClick(position);
        }
      }
    });
  }

  public void setSendBtnListener(OnClickListener sendBtnListener) {
    this.sendBtnListener = sendBtnListener;
  }

  /**
   * 移除tab
   */
  public void removeTab(int position) {
    tabContainer.removeViewAt(position);
    tabList.remove(position);
  }

  public void selectedTo(int position) {
    scrollTo(position);
    for (int i = 0; i < tabList.size(); i++) {
      if (position == i) {
        tabList.get(i).setBackgroundColor(getResources().getColor(R.color.emojicon_tab_selected));
      } else {
        tabList.get(i).setBackgroundColor(getResources().getColor(R.color.emojicon_tab_nomal));
      }
    }
  }

  private void scrollTo(final int position) {
    int childCount = tabContainer.getChildCount();
    if (position < childCount) {
      scrollView.post(new Runnable() {
        @Override
        public void run() {
          int mScrollX = tabContainer.getScrollX();
          int childX = (int) ViewCompat.getX(tabContainer.getChildAt(position));

          if (childX < mScrollX) {
            scrollView.scrollTo(childX, 0);
            return;
          }

          int childWidth = (int) tabContainer.getChildAt(position).getWidth();
          int hsvWidth = scrollView.getWidth();
          int childRight = childX + childWidth;
          int scrollRight = mScrollX + hsvWidth;

          if (childRight > scrollRight) {
            scrollView.scrollTo(childRight - scrollRight, 0);
            return;
          }
        }
      });
    }
  }


  public void setTabBarItemClickListener(EaseScrollTabBarItemClickListener itemClickListener) {
    this.itemClickListener = itemClickListener;
  }


  public interface EaseScrollTabBarItemClickListener {

    void onItemClick(int position);
  }


}

File: app/src/main/java/me/sunhapper/spcharedittool/emoji/EmojiconScrollTabBar.kt
package me.sunhapper.spcharedittool.emoji

import android.content.Context
import android.support.v4.content.ContextCompat
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.RelativeLayout
import kotlinx.android.synthetic.main.common_emoj_widget_emojicon_tab_bar.view.*
import me.sunhapper.spcharedittool.R
import me.sunhapper.spcharedittool.util.Utils
import java.util.*

class EmojiconScrollTabBar @JvmOverloads constructor(private val mContext: Context, attrs: AttributeSet? = null, defStyle: Int = 0)
    : RelativeLayout(mContext, attrs, defStyle) {


    private val tabList = ArrayList<ImageView>()
    private var itemClickListener: EaseScrollTabBarItemClickListener? = null

    private val tabWidth = 60f

    init {
        LayoutInflater.from(mContext).inflate(R.layout.common_emoj_widget_emojicon_tab_bar, this)
    }

    /**
     * 添加tab
     */
    fun addTab(icon: Int) {
        val tabView = View.inflate(mContext, R.layout.common_emoj_scroll_tab_item, null)
        val imageView: ImageView = tabView.findViewById(R.id.iv_icon)
        imageView.setImageResource(icon)
        val imgParams = LinearLayout.LayoutParams(
                Utils.dp2px(mContext, tabWidth), LayoutParams.MATCH_PARENT)
        imageView.layoutParams = imgParams
        tab_container.addView(tabView)
        tabList.add(imageView)
        val position = tabList.size - 1
        imageView.setOnClickListener {
            itemClickListener?.onItemClick(position)
        }
    }


    /**
     * 移除tab
     */
    fun removeTab(position: Int) {
        tab_container.removeViewAt(position)
        tabList.removeAt(position)
    }

    fun selectedTo(position: Int) {
        scrollTo(position)
        for (i in tabList.indices) {
            if (position == i) {
                tabList[i].setBackgroundColor(ContextCompat.getColor(mContext, R.color.emojicon_tab_selected))
            } else {
                tabList[i].setBackgroundColor(ContextCompat.getColor(mContext, R.color.emojicon_tab_nomal))
            }
        }
    }

    private fun scrollTo(position: Int) {
        val childCount = tab_container.childCount
        if (position < childCount) {
            scroll_view.post(Runnable {
                val mScrollX = tab_container.scrollX
                val childX = tab_container.getChildAt(position).x.toInt()

                if (childX < mScrollX) {
                    scroll_view.scrollTo(childX, 0)
                    return@Runnable
                }

                val childWidth = tab_container.getChildAt(position).width
                val hsvWidth = scroll_view.width
                val childRight = childX + childWidth
                val scrollRight = mScrollX + hsvWidth

                if (childRight > scrollRight) {
                    scroll_view.scrollTo(childRight - scrollRight, 0)
                    return@Runnable
                }
            })
        }
    }


    fun setTabBarItemClickListener(itemClickListener: EaseScrollTabBarItemClickListener) {
        this.itemClickListener = itemClickListener
    }


    interface EaseScrollTabBarItemClickListener {

        fun onItemClick(position: Int)
    }


}
